doc/man/man1/vwrays.1
.\" RCSid "$Id: vwrays.1,v 1.13 2022/04/09 17:18:08 greg Exp $"
.TH VWRAYS 1 1/15/99 RADIANCE
.SH NAME
vwrays - compute rays for a given picture or view
.SH SYNOPSIS
.B vwrays
.B "[ -i -u -f{a|f|d} -c rept | -d ]"
{
.B "view opts .."
|
.B picture
.B [zbuf]
}
.SH DESCRIPTION
.I Vwrays
takes a picture or view specification and computes the ray origin and
direction corresponding to each pixel in the image.
This information may then be passed to
.I rtrace(1)
to perform other calculations.
If a given pixel has no corresponding ray (because it is outside the
legal view boundaries), then six zero values are sent instead.
.PP
The
.I \-i
option may be used to specify desired pixel positions on the standard
input rather than generating all the pixels for a given view.
If the
.I \-u
option is also given, output will be unbuffered.
.PP
The
.I \-f
option may be used to set the record format to something other than the
default ASCII.
Using raw float or double records for example can reduce the time
requirements of transferring and interpreting information in
.I rtrace.
.PP
The
.I \-c
option repeats each pixel the given number of times (default is 1).
This is most useful when sending rays to
.I rcontrib(1)
with the same
.I \-c
setting, providing a much faster way to average pixels over image sets.
The
.I \-pj
and/or
.I \-pd
options should be used to jitter sample postions in most cases.
.PP
View options may be any combination of standard view parameters described
in the
.I rpict(1)
manual page, including input from a view file with the
.I \-vf
option.
Additionally, the target X and Y dimensions may be specified with
.I \-x
and
.I \-y
options, and the pixel aspect ratio may be given with
.I \-pa.
The default dimensions are 512x512, with a pixel aspect ratio of 1.0.
Just as in
.I rpict,
the X or the Y dimension will be reduced if necessary
to best match the specified pixel
aspect ratio, unless this ratio is set to zero.
The
.I \-pj
option may be used to jitter samples.
The
.I \-pd
option specifies a world-diameter aperture for depth-of-field jittering.
The default value of 0 for both options turns off all ray jittering.
.PP
If the
.I \-d
option is given, then
.I vwrays
just prints the computed image dimensions, which are based on the view
aspect and the pixel aspect ratio just described.
The
.I \-ld
switch will also be printed, with
.I \-ld+
if the view has an aft clipping plane.
This is useful for passing options to the
.I rtrace
command line.
(See below.)
.PP
If the view contains an aft clipping plane
.I (-va
option), then the magnitudes of the ray directions will
equal the maximum distance for each pixel, which will be interpreted
correctly by
.I rtrace
with the
.I \-ld+
option.
Note that this option should not be given unless there is an aft
clipping plane, since the ray direction vectors will be normalized
otherwise, which would produce a uniform clipping distance of 1.
.PP
If a picture is given on the command line rather than a set of view options,
then the view and image dimensions are taken from the picture file, and
the reported ray origins and directions will match the center of each
pixel in the picture (plus optional jitter).
.PP
If a depth buffer file is given as well, then
.I vwrays
computes the intersection point of each pixel ray (equal to the ray origin
plus the depth times the ray direction), and reports this instead of the
ray origin.
The reported ray direction will also be reversed.
The interpretation of this data is an image of origins and directions
for light rays leaving the scene surfaces to strike each pixel.
.SH EXAMPLES
To compute the ray intersection points and returned directions corresponding
to a picture and its depth buffer:
.IP "" .2i
vwrays scene_v2.hdr scene_v2.zbf > scene_v2.pts
.PP
To determine what the dimensions of a given view would be:
.IP "" .2i
vwrays \-d \-vf myview.vf \-x 2048 \-y 2048
.PP
To generate a RADIANCE picture using
.I rtrace
instead of
.I rpict:
.IP "" .2i
vwrays \-ff \-vf view1.vf \-x 1024 \-y 1024 |
rtrace `vwrays \-d \-vf view1.vf \-x 1024 \-y 1024` \-ffc scene.oct > view1.hdr
.SH AUTHOR
Greg Ward Larson
.SH ACKNOWLEDGMENT
This work was supported by Silicon Graphics, Inc.
.SH BUGS
Although
.I vwrays
can reproduce any pixel ordering (i.e., any image orientation) when given
a rendered picture, it will only produce standard scanline-ordered rays when 
given a set of view parameters.
.SH "SEE ALSO"
rcalc(1), rcode_depth(1), rcontrib(1), rpict(1), rtpict(1), rtrace(1)


src/util/vwrays.c
#ifndef lint
static const char	RCSid[] = "$Id: vwrays.c,v 3.24 2022/04/09 17:18:08 greg Exp $";
#endif
/*
 * Compute rays corresponding to a given picture or view.
 */

#include "platform.h"
#include "standard.h"
#include "random.h"
#include "view.h"

typedef void putfunc(FVECT ro, FVECT rd);
static putfunc puta;
static putfunc putf;
static putfunc putd;
static void pix2rays(FILE *fp);
static void putrays(void);

static putfunc *putr = puta;

VIEW	vw = STDVIEW;

RESOLU	rs = {PIXSTANDARD, 512, 512};

double	pa = 1.;

double  pj = 0.;

double  pd = 0.;

int	zfd = -1;

int	fromstdin = 0;

int	unbuffered = 0;

int	repeatcnt = 1;

char	*progname;


int
main(
	int	argc,
	char	*argv[]
)
{
	char	*err;
	int	rval, getdim = 0;
	int	i;

	progname = argv[0];
	if (argc < 2)
		goto userr;
	for (i = 1; i < argc && argv[i][0] == '-'; i++)
		switch (argv[i][1]) {
		case 'f':			/* output format */
			switch (argv[i][2]) {
			case 'a':			/* ASCII */
				putr = puta;
				break;
			case 'f':			/* float */
				putr = putf;
				SET_FILE_BINARY(stdout);
				break;
			case 'd':			/* double */
				putr = putd;
				SET_FILE_BINARY(stdout);
				break;
			default:
				goto userr;
			}
			break;
		case 'v':			/* view file or option */
			if (argv[i][2] == 'f') {
				rval = viewfile(argv[++i], &vw, NULL);
				if (rval <= 0) {
					fprintf(stderr,
						"%s: no view in file\n",
							argv[i]);
					return(1);
				}
				break;
			}
			rval = getviewopt(&vw, argc-i, argv+i);
			if (rval < 0)
				goto userr;
			i += rval;
			break;
		case 'd':			/* report dimensions only */
			getdim++;
			break;
		case 'x':			/* x resolution */
			rs.xr = atoi(argv[++i]);
			if (rs.xr <= 0) {
				fprintf(stderr, "%s: bad x resolution\n",
						progname);
				return(1);
			}
			break;
		case 'y':			/* y resolution */
			rs.yr = atoi(argv[++i]);
			if (rs.yr <= 0) {
				fprintf(stderr, "%s: bad y resolution\n",
						progname);
				return(1);
			}
			break;
		case 'c':			/* repeat count */
			repeatcnt = atoi(argv[++i]);
			break;
		case 'p':			/* pixel aspect or jitter */
			if (argv[i][2] == 'a')
				pa = atof(argv[++i]);
			else if (argv[i][2] == 'j')
				pj = atof(argv[++i]);
			else if (argv[i][2] == 'd')
				pd = atof(argv[++i]);
			else
				goto userr;
			break;
		case 'i':			/* get pixels from stdin */
			fromstdin = 1;
			break;
		case 'u':			/* unbuffered output */
			unbuffered = 1;
			break;
		default:
			goto userr;
		}
	if ((i > argc) | (i+2 < argc))
		goto userr;
	if (i < argc) {
		rval = viewfile(argv[i], &vw, &rs);
		if (rval <= 0) {
			fprintf(stderr, "%s: no view in picture\n", argv[i]);
			return(1);
		}
		if (!getdim & (i+1 < argc)) {
			zfd = open_float_depth(argv[i+1], (long)rs.xr*rs.yr);
			if (zfd < 0)
				return(1);
		}
	}
	if ((err = setview(&vw)) != NULL) {
		fprintf(stderr, "%s: %s\n", progname, err);
		return(1);
	}
	if (i == argc)
		normaspect(viewaspect(&vw), &pa, &rs.xr, &rs.yr);
	if (getdim) {
		printf("-x %d -y %d%s\n", rs.xr, rs.yr,
				(vw.vaft > FTINY) ? " -ld+" : "");
		return(0);
	}
	if (fromstdin)
		pix2rays(stdin);
	else
		putrays();
	return(0);
userr:
	fprintf(stderr,
	"Usage: %s [ -i -u -f{a|f|d} -c rept | -d ] { view opts .. | picture [zbuf] }\n",
			progname);
	return(1);
}


static void
jitterloc(
	RREAL	loc[2]
)
{
	if (pj > FTINY) {
		loc[0] += pj*(.5 - frandom())/rs.xr;
		loc[1] += pj*(.5 - frandom())/rs.yr;
	}
}


static void
pix2rays(
	FILE *fp
)
{
	static FVECT	rorg, rdir;
	float	zval;
	double	px, py;
	RREAL	loc[2];
	int	pp[2];
	double	d;
	int	i, c;

	while (fscanf(fp, "%lf %lf", &px, &py) == 2) {
		px += .5; py += .5;
		loc[0] = px/rs.xr; loc[1] = py/rs.yr;
		if (zfd >= 0) {
			if ((loc[0] < 0) | (loc[0] >= 1) |
					(loc[1] < 0) | (loc[1] >= 1)) {
				fprintf(stderr, "%s: input pixel outside image\n",
						progname);
				exit(1);
			}
			loc2pix(pp, &rs, loc[0], loc[1]);
			if (lseek(zfd,
				(pp[1]*scanlen(&rs)+pp[0])*sizeof(float),
						SEEK_SET) < 0 ||
					read(zfd, &zval, sizeof(float))
						< sizeof(float)) {
				fprintf(stderr, "%s: depth buffer read error\n",
						progname);
				exit(1);
			}
		}
		for (c = repeatcnt; c-- > 0; ) {
			jitterloc(loc);
			d = viewray(rorg, rdir, &vw, loc[0], loc[1]);
			if (d < -FTINY || !jitteraperture(rorg, rdir, &vw, pd))
				rorg[0] = rorg[1] = rorg[2] =
				rdir[0] = rdir[1] = rdir[2] = 0.;
			else if (zfd >= 0)
				for (i = 0; i < 3; i++) {
					rorg[i] += rdir[i]*zval;
					rdir[i] = -rdir[i];
				}
			else if (d > FTINY) {
				rdir[0] *= d; rdir[1] *= d; rdir[2] *= d;
			}
			(*putr)(rorg, rdir);
			if (c) {
				loc[0] = px/rs.xr; loc[1] = py/rs.yr;
			}
		}
		if (unbuffered)
			fflush(stdout);
	}
	if (!feof(fp)) {
		fprintf(stderr, "%s: expected px py on input\n", progname);
		exit(1);
	}
}


static void
putrays(void)
{
	RREAL	loc[2];
	FVECT	rorg, rdir;
	float	*zbuf = NULL;
	int	sc;
	double	d;
	int	si, i, c;

	if (zfd >= 0) {
		zbuf = (float *)malloc(scanlen(&rs)*sizeof(float));
		if (zbuf == NULL) {
			fprintf(stderr, "%s: not enough memory\n", progname);
			exit(1);
		}
	}
	for (sc = 0; sc < numscans(&rs); sc++) {
		if (zfd >= 0) {
			if (read(zfd, zbuf, scanlen(&rs)*sizeof(float)) <
					scanlen(&rs)*sizeof(float)) {
				fprintf(stderr, "%s: depth buffer read error\n",
						progname);
				exit(1);
			}
		}
		for (si = 0; si < scanlen(&rs); si++) {
		    for (c = repeatcnt; c-- > 0; ) {
			pix2loc(loc, &rs, si, sc);
			jitterloc(loc);
			d = viewray(rorg, rdir, &vw, loc[0], loc[1]);
			if (d < -FTINY || !jitteraperture(rorg, rdir, &vw, pd))
				rorg[0] = rorg[1] = rorg[2] =
				rdir[0] = rdir[1] = rdir[2] = 0.;
			else if (zfd >= 0)
				for (i = 0; i < 3; i++) {
					rdir[i] = -rdir[i]*zbuf[si];
					rorg[i] -= rdir[i];
				}
			else if (d > FTINY) {
				rdir[0] *= d; rdir[1] *= d; rdir[2] *= d;
			}
			(*putr)(rorg, rdir);
		    }
		}
	}
	if (zfd >= 0)
		free((void *)zbuf);
}


static void
puta(		/* put out ray in ASCII format */
	FVECT	ro,
	FVECT	rd
)
{
	printf("%.5e %.5e %.5e %.5e %.5e %.5e\n",
			ro[0], ro[1], ro[2],
			rd[0], rd[1], rd[2]);
}


static void
putf(		/* put out ray in float format */
	FVECT	ro,
	FVECT	rd
)
{
	float v[6];

	v[0] = ro[0]; v[1] = ro[1]; v[2] = ro[2];
	v[3] = rd[0]; v[4] = rd[1]; v[5] = rd[2];
	putbinary(v, sizeof(float), 6, stdout);
}


static void
putd(		/* put out ray in double format */
	FVECT	ro,
	FVECT	rd
)
{
	double v[6];

	v[0] = ro[0]; v[1] = ro[1]; v[2] = ro[2];
	v[3] = rd[0]; v[4] = rd[1]; v[5] = rd[2];
	putbinary(v, sizeof(double), 6, stdout);
}
